static/app/components/forms/teamSelector.tsx

```diff
@@ -0,0 +1,264 @@
+   */
+  teamFilter?: (team: Team) => boolean;
+  /**
+   * Can be used to restrict teams to a certain project and allow for new teams to be add to that project
+   */
+  project?: Project;
+  /**
+   * Controls whether the value in the dropdown is a team id or team slug
+   */
+  useId?: boolean;
+  includeUnassigned?: boolean;
+} & ControlProps;
+
+type TeamActor = {
+  type: 'team';
+  id: string;
+  name: string;
+};
+
+type SelectableTeam = {
+  value: string;
+  label: React.ReactElement;
+  searchKey: string;
+  actor: TeamActor;
+  disabled?: boolean;
+};
+
+type TeamOption = SelectableTeam | typeof unassignedOption;
+
+type State = {
+  options: TeamOption[];
+};
+
+class TeamSelector extends React.Component<Props, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      options: this.teamOptions,
+    };
+  }
+
+  // TODO(ts) This type could be improved when react-select types are better.
+  selectRef = React.createRef<any>();
+
+  get teamOptions() {
+    const {teams, teamFilter, includeUnassigned, project} = this.props;
+    const filteredTeams = teamFilter ? teams.filter(teamFilter) : teams;
+    if (project) {
+      const teamsInProjectIdSet = new Set(project.teams.map(({id}) => id));
+      const teamsInProject = filteredTeams.filter(({id}) => teamsInProjectIdSet.has(id));
+      const teamsNotInProject = filteredTeams.filter(
+        ({id}) => !teamsInProjectIdSet.has(id)
+      );
+
+      return [
+        ...teamsInProject.map(this.createTeamOption),
+        ...teamsNotInProject.map(this.createTeamOutsideProjectOption),
+        ...(includeUnassigned ? [unassignedOption] : []),
+      ];
+    }
+    const teamOptions: TeamOption[] = filteredTeams.map(this.createTeamOption);
+    if (includeUnassigned) {
+      teamOptions.push(unassignedOption);
+    }
+    return teamOptions;
+  }
+
+  createTeamOption = (team: Team): SelectableTeam => ({
+    value: this.props.useId ? team.id : team.slug,
+    label: <IdBadge team={team} />,
+    searchKey: `#${team.slug}`,
+    actor: {
+      type: 'team',
+      id: team.id,
+      name: team.slug,
+    },
+  });
+
+  /**
+   * Closes the select menu by blurring input if possible since that seems to be the only
+   * way to close it.
+   */
+  closeSelectMenu() {
+    if (!this.selectRef.current) {
+      return;
+    }
+
+    const select = this.selectRef.current.select;
+    const input: HTMLInputElement = select.inputRef;
+    if (input) {
+      // I don't think there's another way to close `react-select`
+      input.blur();
+    }
+  }
+
+  handleAddTeamToProject = async (team: Team) => {
+    const {api, organization, project, value} = this.props;
+    const {options} = this.state;
+
+    // Copy old value
+    const oldValue = value ? [...value] : {value};
+
+    // Optimistic update
+    this.props.onChange?.(this.createTeamOption(team));
+
+    try {
+      // Try to add team to project
+      if (project) {
```

Comment:
Should we early exit outside of this try catch to keep the control flow logic simpler.